File: scripts/verify-ability-scaling.mjs
/**
 * Fetch official wiki wikitext for abilities in the scaling registry
 * and extract stat / scaling hints for manual verification.
 */
import { writeFileSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const API = "https://wiki.warframe.com/api.php";
const HEADERS = { "User-Agent": "Voidforge/1.0 (https://void-forge.org)" };

const PAGES = [
  "Gravitic Slash",
  "Astral Shell",
  "Light's Sanctuary",
  "Event Horizon",
  "Shatter Shield",
  "Peacemaker",
  "Virulence",
  "Parasitic Link",
  "Ravenous",
  "Iron Skin",
  "Molt",
  "Toxic Lash",
  "Tharros Strike",
  "Rally Point",
  "Final Stand",
  "Speed",
  "Electric Shield",
  "Cloud Walker",
  "Defy",
  "Gloom",
  "Eclipse",
  "Self Portrait",
  "Plein Air",
  "Psychic Bolts",
];

async function fetchWikitext(title) {
  const params = new URLSearchParams({
    action: "parse",
    page: title,
    prop: "wikitext",
    format: "json",
  });
  const resp = await fetch(`${API}?${params}`, { headers: HEADERS });
  if (!resp.ok) return { error: resp.statusText };
  const data = await resp.json();
  if (data.error) return { error: data.error.info };
  return { wikitext: data.parse?.wikitext?.["*"] ?? "" };
}

function extractScalingHints(wikitext) {
  const hints = [];
  const patterns = [
    /Ability Strength/gi,
    /Ability Duration/gi,
    /Ability Range/gi,
    /Ability Efficiency/gi,
    /\|STR\|/g,
    /\|DUR\|/g,
    /\|RNG\|/g,
    /cap(?:ped)? at \d+%/gi,
    /maximum of \d+%/gi,
    /scales with/gi,
    /does not scale/gi,
    /not affected by/gi,
  ];
  for (const pat of patterns) {
    const matches = wikitext.match(pat);
    if (matches) hints.push(...matches.map((m) => m.toLowerCase()));
  }
  // Infobox stat rows (common template format)
  const statRows = [...wikitext.matchAll(/\|\s*([^\n|]+?)\s*=\s*([^\n|]+)/g)]
    .slice(0, 40)
    .map((m) => `${m[1].trim()}: ${m[2].trim()}`);
  return { hints: [...new Set(hints)], statRows };
}

const results = {};

for (const page of PAGES) {
  process.stderr.write(`Fetching ${page}...\n`);
  const { wikitext, error } = await fetchWikitext(page);
  if (error) {
    results[page] = { error };
  } else {
    const { hints, statRows } = extractScalingHints(wikitext);
    results[page] = {
      hints,
      statRows: statRows.filter((r) =>
        /strength|duration|range|strip|reduction|radius|armor|shield|damage|slow|heal|regen|cap/i.test(r),
      ),
      excerpt: wikitext.slice(0, 2500),
    };
  }
  await new Promise((r) => setTimeout(r, 350));
}

const outPath = join(__dirname, "..", "tmp-wiki-ability-verify.json");
writeFileSync(outPath, JSON.stringify(results, null, 2));
console.log(`Wrote ${outPath}`);

File: scripts/verify-ability-scaling.py
#!/usr/bin/env python3
"""
Fetch official wiki.warframe.com wikitext for abilities in the scaling registry.

Run from repo root:
  python scripts/verify-ability-scaling.py

Output: tmp-wiki-ability-verify.json (gitignored) — use to audit
src/lib/ability-scaling-registry.ts against wiki Ability templates.
"""
import json
import re
import time
import urllib.parse
import urllib.request

API = "https://wiki.warframe.com/api.php"
HEADERS = {"User-Agent": "Voidforge/1.0 (https://void-forge.org)"}

PAGES = [
    "Gravitic Slash", "Astral Shell", "Light's Sanctuary", "Event Horizon",
    "Shatter Shield", "Peacemaker", "Virulence", "Parasitic Link", "Ravenous",
    "Iron Skin", "Molt", "Toxic Lash", "Tharros Strike", "Rally Point", "Final Stand",
    "Speed", "Electric Shield", "Cloud Walker", "Defy", "Gloom", "Eclipse",
    "Self Portrait", "Plein Air", "Psychic Bolts",
]


def fetch_wikitext(title: str) -> str:
    params = urllib.parse.urlencode({
        "action": "parse",
        "page": title,
        "prop": "wikitext",
        "format": "json",
    })
    req = urllib.request.Request(f"{API}?{params}", headers=HEADERS)
    with urllib.request.urlopen(req, timeout=30) as resp:
        data = json.loads(resp.read().decode())
    if "error" in data:
        raise RuntimeError(data["error"].get("info", str(data["error"])))
    return data["parse"]["wikitext"]["*"]


def main():
    results = {}
    for page in PAGES:
        print(f"Fetching {page}...", flush=True)
        try:
            wt = fetch_wikitext(page)
            # Pull Ability template stat lines
            stat_lines = re.findall(
                r"\|([^|\n=]+?)\s*=\s*([^\n|]+)",
                wt,
            )
            relevant = [
                f"{k.strip()}: {v.strip()}"
                for k, v in stat_lines
                if re.search(
                    r"strength|duration|range|strip|reduction|radius|armor|shield|"
                    r"damage|slow|heal|regen|cap|bonus|multiplier|arc|width|javelin|"
                    r"maggot|energy|status|explosion|decoy|sanctuary|horizon",
                    k + v,
                    re.I,
                )
            ]
            results[page] = {
                "relevantStats": relevant[:60],
                "scalingNotes": [
                    m.group(0)
                    for m in re.finditer(
                        r"(?i)(ability strength|ability duration|ability range|"
                        r"scales with|does not scale|not affected by|cap(?:ped)? at \d+%|"
                        r"maximum of \d+%)",
                        wt,
                    )
                ][:30],
                "excerpt": wt[:3000],
            }
        except Exception as e:
            results[page] = {"error": str(e)}
        time.sleep(0.35)

    out = "tmp-wiki-ability-verify.json"
    with open(out, "w", encoding="utf-8") as f:
        json.dump(results, f, indent=2, ensure_ascii=False)
    print(f"Wrote {out}")


if __name__ == "__main__":
    main()
